trl/trainer/ppo_trainer.py

```diff
@@ -649,76 +662,82 @@ def step(
             "logprobs": all_logprobs.to(torch.float32),
             "values": values.to(torch.float32),
-            "rewards": rewards,
             "masks": masks,
+            "advantages": advantages,
+            "returns": returns,
         }
-
-        def collator(data):
-            return_dict = dict()
-            for key in data[0]:
-                if key in ["queries", "responses"]:
-                    return_dict[key] = [d[key] for d in data]
-                else:
-                    return_dict[key] = torch.stack([d[key] for d in data]).to(self.current_device)
-            return return_dict
-
-        mini_batch_dict.update(model_inputs)
-        mini_batch_data = Dataset.from_dict(mini_batch_dict)
-        mini_batch_data.set_format("torch")
-        mini_batch_dataloader = torch.utils.data.DataLoader(
-            mini_batch_data,
-            batch_size=self.config.mini_batch_size,
-            shuffle=True,
-            collate_fn=collator,
-        )
+        batch_dict.update(model_inputs)
 
         t = time.time()
         all_stats = []
         early_stop = False
         for _ in range(self.config.ppo_epochs):
             if early_stop:
                 break
-
-            for i, batch in enumerate(mini_batch_dataloader):
-                with self.accelerator.accumulate(self.model):
-                    model_inputs = {k: batch[k] for k in model_inputs_names}
-                    logprobs, logits, vpreds, _ = self.batched_forward_pass(
-                        self.model,
-                        batch["queries"],
-                        batch["responses"],
-                        model_inputs,
-                        return_logits=True,
-                    )
-                    if (i % self.config.gradient_accumulation_steps) == 0:
-                        self.optimizer.zero_grad()
-
-                    train_stats = self.train_minibatch(
-                        batch["logprobs"],
-                        batch["values"],
-                        batch["rewards"],
-                        logprobs,
-                        logits,
-                        vpreds,
-                        batch["masks"],
-                    )
-
-                    all_stats.append(train_stats)
-
-                    if self.config.early_stopping:
-                        policykl = train_stats["policy/policykl"]
-                        early_stop = self._early_stop(policykl)
-                        if early_stop:
-                            break
+            b_inds = np.random.permutation(bs)
+            for mini_batch_start in range(0, bs, self.config.mini_batch_size):
+                mini_batch_end = mini_batch_start + self.config.mini_batch_size
+                mini_batch_inds = b_inds[mini_batch_start:mini_batch_end]
+
+                # set optimizer to zero for gradient accumulation
+                for micro_batch_start in range(0, self.config.mini_batch_size, self.config.micro_batch_size):
```

Comment:
Hi, 

I'm trying to understand the motivation behind the introduction of `micro_batch` in addition to `mini_batch`.

For example, given the contrived PPO config:
```
mini_batch_size = 4
gradient_accumulation_steps = 2
micro_batch_size = mini_batch_size / gradient_accumulation_steps = 2
```
(A side question: why even make `micro_batch_size` configurable when you will overwrite it with `mini_batch_size / gradient_accumulation_steps`?)

Based on my understanding:
Original:
- Train a batch of 4 (`mini_batch_size`), and zero out gradients for every 2 (`gradient_accumulation_steps`) batches;

New:
- Train a batch of 2 (`micro_batch_size`), and always zero out gradients;
- This is equivalent to setting `mini_batch_size = 2` and `gradient_accumulation_steps = 1` in the original implementation.

So it looks like
1. The behaviors between the original and new implementations are different based on the same `mini_batch_size` and `gradient_accumulation_steps` config values, which could make this change not backward compatible;
2. The additional level of micro batch for loop appears to be unnecessary and further increase code complexity when you can simply use different `mini_batch_size` and `gradient_accumulation_steps` config values to achieve the same behavior;
3. `gradient_accumulation_steps` in the new implementation loses its intended meaning because we do not accumulate gradients at all with micro batches.

Thanks and sorry if I misunderstood anything,

Felix

Reply:
Hi Felix, thanks for the thoughtful reply. 

> A side question: why even make micro_batch_size configurable when you will overwrite it with mini_batch_size / gradient_accumulation_steps?

Because then `micro_batch_size` will be recorded to wandb for experiment tracking to my understanding. 

> Train a batch of 2 (micro_batch_size), and always zero out gradients;... gradient_accumulation_steps in the new implementation loses its intended meaning because we do not accumulate gradients at all with micro batches.

It looks this way but in reality, it's doing something different. Because we hand off gradient accumulation to accelerate, you are supposed to call `optimizer.zero_grad()` even though typically you wouldn't (see https://huggingface.co/docs/accelerate/usage_guides/gradient_accumulation#the-finished-code)

> The additional level of micro batch for loop appears to be unnecessary and further increase code complexity when you can simply use different mini_batch_size and gradient_accumulation_steps config values to achieve the same behavior;

I guess this is more of a stylistic thing. We really have three levels of "batch sizes":

1. `batch_size` is the amount of rollout data, e,g., 8 data points
2. `minibatch_size` is the amount of data that you are actually doing a  `zero_grad()`, `loss.backward()`, and `optimizer.step()`, e.g., 4 data points
3. `microbatch_size` if the `minibatch_size` is too large to fit in memory, so we `zero_grad()`, partition the `minibatch` and do multiple `loss.backward()` and then `optimizer.step()`. 

Having the terminology like this makes it clear what the real `minibatch_size` is, so that we do not confound it with `gradient_accumulation_steps`. 

From the user's perspective, one set of hyperparameters should always reliably reproduce similar learning curves. By the new API design in this PR, `batch_size=8, minibatch_size=4` will reliably reproduce similar learning curves regardless of how many `gradient_accumulation_steps` there is, so when doing experiment management / analysis we can just group /filter by `batch_size` and `minibatch_size`. The existing implementation, however, would require us to calculate what's the "real size of the data we perform an `optimizer.step()` on".